### PR TITLE
Add support for custom workflows

### PR DESCRIPTION
**Pull request by DominicKramer**
Adds support for performing work that needs to be done to get the
application directory in a state ready for starting the
application.  Such work could include, for example, running gulp,
typescript, or webpack.

### CODE REVIEW

File: builder/steps/gen-dockerfile/contents/data/Dockerfile.txt [hunk 1]
```diff
@@ -1,14 +1,52 @@
...
+# This command will also cat the npm-debug.log file after the
+# build, if it exists.
+RUN npm install --unsafe-perm || \
+  ((if [ -f npm-debug.log ]; then \
+      cat npm-debug.log; \
+    fi) && false)
+  <% } %>
+ENV NODE_ENV production
+<% } %>
+RUN <%- config.buildCommand %>
```
Comment (duggelz) [inline builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: For security/paranoia's sake, should you check that "config.buildCommand" doesn't have a newline or similar?  Basically I'm thinking the Bash/Docker equivalent of SQL injection attacks, like someone's app.yaml having `runtime_config: build: "true\\nCOPY /etc/passwd /pwned"`.

Comment (ofrobots) [inline builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: If an attacker controls your `app.yaml`, they could also do this: `runtime_config: build: COPY /etc/passwd /pwned`. So checking for newline doesn't really decrease that attack.. or am I missing something?

Comment (duggelz) [inline builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: `RUN cp /etc/passwd /pwned` would get the /etc/passwd from inside the container, whereas `COPY /etc/passwd /pwned` would let the user reach out to the Argo build VM, and copy arbitrary stuff into the container for later perusal.  I don't know what specifically is sensitive inside the build VM, but I haven't seen a document that says "0% guaranteed, it's fine" either.

If nothing else, we don't want users to realize there is a "feature" of "Can inject arbitrary Dockerfile directives" and then start using and depending on that "feature"

Comment (DominicKramer) [inline builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: @duggelz I see your point, but the user could also just use their own Dockerfile if they wanted and then do whatever they want.  That is, we should make sure what you are addressing is not an issue.  However, if it is, its scope is beyond just this PR.

Comment (duggelz) [inline builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: That's a good point about customer Dockerfiles.  Ok.

**Comment by ofrobots** [comment]
@jmdobry I wonder how you managed to remove the 'cla: yes' label? Accident? Buggy automation?

**Comment by jmdobry** [comment]
@ofrobots I have no idea, I don't even look at this repo.

File: integration/test/test.ts [hunk 2]
```diff
@@ -76,10 +76,20 @@ const CONFIGURATIONS: TestConfig[] = [
...
+    expectedRunStderr: '',
+    buildStdoutContains: [
+      `FROM ${RUNTIME_TAG} as build_step`, 'RUN npm run gcp-build',
+      'FROM build_step', 'COPY --from=build_step'
+    ]
   }
 ];
 
-const DEBUG = false;
+const DEBUG = true;
```
Comment (ofrobots) [inline integration/test/test.ts]: intentional?

File: builder/steps/gen-dockerfile/contents/data/Dockerfile.txt [hunk 3]
```diff
@@ -1,6 +1,40 @@
...
+    fi) && false)
+  <% } %>
+<% } %>
+RUN npm run gcp-build
+  <% if (config.canInstallDeps) { %>
+RUN rm -Rf node_modules
+  <% } %>
+<% } %>
+
+<% if (config.hasBuildCommand) { %>
```
Comment (ofrobots) [inline builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: as discussed, this if can be merged with the previous if body

File: builder/steps/gen-dockerfile/contents/data/Dockerfile.txt [hunk 4]
```diff
@@ -25,7 +59,9 @@ RUN yarn global add npm@'<%- config.npmVersion %>'
 RUN yarn global add yarn@'<%- config.yarnVersion %>'
 <% } %>
 
+<% if (!config.hasBuildCommand) { %>
```
Comment (ofrobots) [inline builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: As discussed, this if can be avoided if you move the COPY up to the else block above.

Comment (DominicKramer) [reply builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: I've moved it.

File: builder/steps/gen-dockerfile/contents/data/Dockerfile.txt (hunk 3 above)
Comment (DominicKramer) [reply builder/steps/gen-dockerfile/contents/data/Dockerfile.txt]: Updated

File: integration/test/test.ts (hunk 2 above)
Comment (DominicKramer) [reply integration/test/test.ts]: No.  I switched it back.

**Pull request merged**